### ISSUE TITLE
`Credo.Check.Readability.Specs` fails to find the spec for `template_not_found/2` in Phoenix

### ISSUE THREAD

**Issue by bmitc** [rrrene/credo#984]
### Environment

* Credo version (`mix credo -v`): `1.6.4-ref.main.6240794+uncommittedchanges`
* Erlang/Elixir version (`elixir -v`):
    ```
    Erlang/OTP 24 [erts-12.1.4] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit]

    Elixir 1.13.3 (compiled with Erlang/OTP 22)
    ```
* Operating system: Windows 11 and Ubuntu via WSL2

### What were you trying to do?
Trying to get the Credo check `Credo.Check.Readability.Specs` to pass on a Phoenix generated project.

### Expected outcome
Credo passes.

### Actual outcome
`Credo.Check.Readability.Specs` does not pass. The issue may be due to the heavy use of macros in Phoenix or the use of `defoverridable` where `template_not_found/2` is defined.

---

Phoenix generates the following module for a fresh Mix and Phoenix project named `islands_interface`, created with `mix phx.new islands_interface --no-ecto`:

```elixir
defmodule IslandsInterfaceWeb.ErrorView do
  use IslandsInterfaceWeb, :view

  def template_not_found(template, _assigns) do
    Phoenix.Controller.status_message_from_template(template)
  end
end
```

I implement Credo in my codebases and run it in strict mode. The function `template_not_found/2` fails the `Credo.Check.Readability.Specs` Credo check, despite there being a [typespec defined for it in the Phoenix.Template module in Phoenix.View](https://github.com/phoenixframework/phoenix_view/blob/master/lib/phoenix/template.ex#L148-L159).

If I add any `@spec` to `IslandsInterfaceWeb.ErrorView.template_not_found/2`, including the exact spec listed in the link above, then I get the following Dialyzer error:

```
Overloaded contract for IslandsInterfaceWeb.ErrorView.template_not_found/2 has
overlapping domains; such contracts are currently unsupported and
are simply ignored.
```

There is more information here: https://elixirforum.com/t/issues-with-credo-dialyzer-and-compiler-warnings-for-app-name-web-errorview-template-not-found-2-in-phoenix/48877

**Comment by rrrene** [comment]
This is a limitation of the check, which is why this check is tagged as "experimental". Maybe we should display some kind of disclaimer when an experimental check raises an issue 🤔 

I am closing this issue because it is a duplicate of https://github.com/rrrene/credo/issues/964#issuecomment-1099570210

**Comment by bmitc** [comment]
Thanks for the update @rrrene! I apologize for not seeing the existing ticket. Believe or not, I did perform a search, and I'm almost certain I searched for "defoverridable".

I'll take a look at investigating this since it's confirmed to be an issue with the check. I also think there's a possibility that `template_not_found/2` is not defined properly as a callback in Phoenix, but that is a separate issue.

**Comment by rrrene** [comment]
@bmitc You are very much welcome to look for a fix! The main problem is that Credo is a *static* code analysis tool and thus has to interpret the source code by reading it, i.e. Credo can not compile the code and look at the `.beam` files.

**Comment by zoten** [comment]
Good morning!
I was going to open an issue too but I feel like the problem is very similar here (altough different).
I showed it in [this](https://gist.github.com/zoten/623d46d20f2ae4d3a48f46bd59d2ca1b) gist, where I define some functions in a macro with dyamic names and even if a spec is created too `credo` complains for it not existing (tried other ways of creating the spec, as suggested [here](https://elixirforum.com/t/credo-and-specs-for-macro-generated-functions/49196/2) with no luck )
Do you think @rrrene this is worth a different issue, or do we make it fall under the "static analyzer" limitations?
As always, thanks for a wonderful work :pray:

**Comment by rrrene** [comment]
@zoten  Yeah, this sounds like it's the same issue :+1:
